src/compiler/scala/tools/nsc/ast/TreeDSL.scala

```diff
@@ -140,7 +140,18 @@ trait TreeDSL {
     def NEW(tpt: Tree, args: Tree*): Tree   = New(tpt, List(args.toList))
 
     def NOT(tree: Tree)   = Select(tree, Boolean_not)
-    def AND(guards: Tree*) = if (guards.isEmpty) EmptyTree else guards reduceLeft gen.mkAnd
+    def AND(guards: Tree*) = {
+      def binaryTreeAnd(tests: Seq[Tree]): Tree = tests match{
```

Comment:
I think it is better to require List here and recurse on `Nil, a :: Nil, twoOrMore` convert to List before the outer call.

Also you could have this method be:

```scala
def binaryTreeAnd(head: Tree, tail: List[Tree]): Tree
```

And then this method clearly never has nested EmptyTree.

The EmptyTree case can be handled once at the outer level of the input is empty.

Reply:
Actually the use of splitAt recursively would make List version quadratic. I don't know if ArraySeq has O(1) splitAt. If it doesn't, the current code is also quadratic.

A better approach might be to copy to an Array and recurse on indices representing the upper and lower bounds of the array which will certainly only N log N.

Reply:
btw a quick macro using `AND()` shows the empty tree is emitted as false in the end, the default value. You might expect zero predicates to be trivially true instead.
```
def f(): Boolean = (<empty>: Boolean);
```

Reply:
@johnynek varargs are instantiated as `ArraySeq`. `splitAt` for that class seems to be implemented by s.c.AbstractIterable, which I presume uses take + drop as implementation. I'd probably go with getting the array from the ArraySeq and work with offsets, as you suggested, as then you can avoid allocating tuples and new collections.